modules/SerialCommHub/tiny_modbus_rtu.cpp

```diff
@@ -0,0 +1,327 @@
+        timeout.tv_usec = MODBUS_RX_WITHIN_MESSAGE_TIMEOUT_MS *
+                          1000; // reduce timeout after first chunk, no uneccesary waiting at the end of the message
+        if (rv == -1) {         // error in select function call
+            perror("txrx: select:");
+            break;
+        } else if (rv == 0) { // no more bytes to read within timeout, so transfer is complete
+            break;
+        } else {              // received more bytes, add them to buffer
+            // do we have space in the rx buffer left?
+            if (bytes_read_total >= rxbuf_len) {
+                // no buffer space left, but more to read.
+                break;
+            }
+
+            int bytes_read = read(fd, rxbuf + bytes_read_total, rxbuf_len - bytes_read_total);
+            if (bytes_read > 0) {
+                bytes_read_total += bytes_read;
+                // EVLOG_info << "RECVD: " << hexdump(rxbuf, bytes_read_total);
+            }
+        }
+    }
+    return bytes_read_total;
+}
+
+/*
+    This function transmits a modbus request and waits for the reply.
+    Parameter request is optional and is only used for writing multiple registers.
+*/
+std::vector<uint16_t> TinyModbusRTU::txrx(uint8_t device_address, FunctionCode function,
+                                          uint16_t first_register_address, uint16_t register_quantity,
+                                          bool wait_for_reply, std::vector<uint16_t> request) {
+    {
+        // size of request
+        int req_len = (request.size() == 0 ? 0 : 2 * request.size() + 1) + MODBUS_BASE_PAYLOAD_SIZE;
+        uint8_t* req = new uint8_t[req_len];
+
+        // add header
+        req[DEVICE_ADDRESS_POS] = device_address;
+        req[FUNCTION_CODE_POS] = function;
+
+        first_register_address = htobe16(first_register_address);
+        register_quantity = htobe16(register_quantity);
+        memcpy(req + REQ_TX_FIRST_REGISTER_ADDR_POS, &first_register_address, 2);
+        memcpy(req + REQ_TX_QUANTITY_POS, &register_quantity, 2);
+
+        if (function == FunctionCode::WRITE_MULTIPLE_HOLDING_REGISTERS) {
+            // write byte count
+            req[REQ_TX_MULTIPLE_REG_BYTE_COUNT_POS] = request.size() * 2;
+            // add request data
+            int i = REQ_TX_MULTIPLE_REG_BYTE_COUNT_POS + 1;
+            for (auto r : request) {
+                r = htobe16(r);
+                memcpy(req + i, &r, 2);
+                i += 2;
+            }
+        }
+        // set cksum in the last 2 bytes
+        append_checksum(req, req_len);
+
+        // EVLOG_info << "SEND: " << hexdump(req, req_len);
+
+        // clear input and output buffer
+        tcflush(fd, TCIOFLUSH);
+
+        // write to serial port
+        write(fd, req, req_len);
+        tcdrain(fd);
+
+        if (ignore_echo) {
+            // read back echo of what we sent and ignore it
+            read_reply(req, req_len);
+        }
+
+        delete req;
```

Comment:
if `new[]`, you need to use `delete[]`.  Interesting that it didn't break already